Add release notes for 2.10.0 (#9119)
PrefectHQ/prefect: RELEASE-NOTES.md

![image](https://user-images.githubusercontent.com/26799928/230421783-997e4fda-a02f-4bf4-88e1-f51a2f890cf5.png)

Configure the prefect client to by default send block method call telemetry to prefect cloud - https://github.com/PrefectHQ/prefect/pull/9107
...

### Versioned documentation
Since we often release new versions of Prefect with new features, we're adding versioned documentation to make it easier to find the docs for the version of Prefect you're using. 

Now, when you visit the Prefect documentation site, you'll see a version selector at the top of the page. 

![versioned docs](https://user-images.githubusercontent.com/228762/230432235-26fc9406-1390-4c63-9956-b8cdabdfba6f.png)

### Breaking Changes

> - Some unsused options for sorting logs have been removed from the API — https://github.com/PrefectHQ/prefect/pull/7873

Reviewer: ```suggestion
- Some unused options for sorting logs have been removed from the API — https://github.com/PrefectHQ/prefect/pull/7873
```

### Workers [Beta]

Workers are next-generation agents, designed from the ground up to interact with [work pools](https://docs.prefect.io/concepts/work-pools/). Each worker manages flow run infrastructure of a specific type and must pull from a work pool with a matching type. Existing work pools are all "agent" typed for backwards compatibility with our agents — but new work pools can be assigned a specific infrastructure type. Specifying a type for a work pool simplifies choosing what kind of infrastructure will be used when creating a flow run. 

> Work pools expose rich configuration of their infrastructure. Every work pool type has a base configuration with sensible defaults such that you can begin executing work with just a single command. The infrastructure configuration is fully customizable from the Prefect UI. For example, you can now customize the entire payload used for run flows on Kubernetes — you are not limited to the fields Prefect exposes in its SDK. We provide templating to inject runtime information and common settings into infrastructure creation payloads. Advanced users can add _custom_ template variables which are then exposed the same as Prefect's default options in an easy to use UI.

Reviewer 2: ```suggestion
Work pools expose rich configuration of their infrastructure. Every work pool type has a base configuration with sensible defaults such that you can begin executing work with just a single command. The infrastructure configuration is fully customizable from the Prefect UI. For example, you can now customize the entire payload used to run flows on Kubernetes — you are not limited to the fields Prefect exposes in its SDK. We provide templating to inject runtime information and common settings into infrastructure creation payloads. Advanced users can add _custom_ template variables which are then exposed the same as Prefect's default options in an easy to use UI.
```

```python
from prefect import variables

# from a synchronous context
answer = variables.get('the_answer')
print(answer)
# 42

# from an asynchronous context
answer = await variables.get('the_answer')
print(answer)
# 42

# without a default value
answer = variables.get('not_the_answer')
print(answer)
# None

# with a default value
answer = variables.get('not_the_answer', default='42')
print(answer)
# 42
```

See the new [variables concept doc](https://docs.prefect.io/latest/concepts/variables/) for more information or the [pull request](https://github.com/PrefectHQ/prefect/pull/9088) for implementation details.

### Events

> Continuing the rollout of (events[https://docs.prefect.io/concepts/events-and-resources/]) as a primary unit of observability in Prefect Cloud, Prefect will now by default emit events for all block method calls. These events can be viewed in the Event feed, allowing you to analyze the interactions your flows and tasks have with external systems such as storage locations, notification services, and infrastructure. Additionally, you can configure automations based on these events, such as when a file is uploaded to a storage location or an infrastructure process or an infrastructure block starts a job.

Reviewer 2: ```suggestion
Continuing the rollout of (events[https://docs.prefect.io/concepts/events-and-resources/]) as the primary unit of observability in Prefect Cloud, Prefect will now emit events for all block method calls by default. These events can be viewed in the Event feed, allowing you to analyze the interactions your flows and tasks have with external systems such as storage locations, notification services, and infrastructure. Additionally, you can trigger automations based on these events. For example, you can create an automation that is triggered when a file is uploaded to a storage location.
```

Reviewer 2: @WillRaphaelson  any concerns with these revisions?

![image](https://user-images.githubusercontent.com/26799928/230421783-997e4fda-a02f-4bf4-88e1-f51a2f890cf5.png)

### Versioned documentation

> We're releasing a lot of new features every week and we know not everyone is on the latest version. We've added versioning to our documentation website to make it easier to find the docs for the version of Prefect you're using.

Reviewer 2: ```suggestion
We're releasing a lot of new features every week, sometimes we substantial changes to behavior, but not everyone is on the latest version of Prefect. We've added versioning to our documentation website to make it easier to find the docs for the version of Prefect that you're using.
```

Author: I'm hesitant to call out substantial changes to behavior — did the rest though.

Reviewer 2: Makes sense

If the work pool’s configuration is updated, all workers automatically begin using the new settings — you no longer need to redeploy your agents to change infrastructure settings. For advanced use cases, you can override settings on a per-deployment basis.

<img width="1601" alt="Creating a new Kubernetes work pool" src="https://user-images.githubusercontent.com/2586601/230471683-63875a04-f331-4cf1-8b1b-69c2cd0e4e05.png">
<img width="1601" alt="Advanced configuration of the work pool infrastructure" src="https://user-images.githubusercontent.com/2586601/230471686-7146e930-34fc-43ae-a946-9e3795c4a27a.png">

Adding a new variable to the advanced work pool configuration will expose it in the basic config:

<img width="1551" alt="Adding a variable to the advanced config" src="https://user-images.githubusercontent.com/2586601/230475075-b535b158-62a8-4b88-9439-0054f58e8f77.png">
<img width="1601" alt="New variables can be adjusted in the basic config" src="https://user-images.githubusercontent.com/2586601/230473701-b8db1973-eb03-4682-86cc-64b698356048.png">

> 

Reviewer 2: ```suggestion

This release includes both Kubernetes and Docker infrastructure types. Additional infrastructure types will be included in subsequent releases.

```

Reviewer 2: @madkinsz any concerns with adding this?

- Update default agent query interval from 10s to 15s — https://github.com/PrefectHQ/prefect/pull/9085
- Add a 10 minute cache to API healthchecks — https://github.com/PrefectHQ/prefect/pull/9069
- Improve performance of concurrent task runner — https://github.com/PrefectHQ/prefect/pull/9073
- Improve performance of waiting for task submission — https://github.com/PrefectHQ/prefect/pull/9072
- Add retry on 502 BAD GATEWAY to client — https://github.com/PrefectHQ/prefect/pull/9102
- Update local and remote file systems to return path on write — https://github.com/PrefectHQ/prefect/pull/8965
- Add artifacts `/count` route — https://github.com/PrefectHQ/prefect/pull/9022
- Improve performance of automatic block registration — https://github.com/PrefectHQ/prefect/pull/8838
- Improve performance of log retrieval queries — https://github.com/PrefectHQ/prefect/pull/9035
- Improve performance of artifact retrieval — https://github.com/PrefectHQ/prefect/pull/9061 / https://github.com/PrefectHQ/prefect/pull/9064
- Add `--type` option to create work-pool CLI — https://github.com/PrefectHQ/prefect/pull/8993

> 

Author: The first one is already included as the first item

```suggestion
- Improve flow run timeline performance - https://github.com/PrefectHQ/prefect-ui-library/pull/1315
```

- Update default agent query interval from 10s to 15s — https://github.com/PrefectHQ/prefect/pull/9085
- Add a 10 minute cache to API healthchecks — https://github.com/PrefectHQ/prefect/pull/9069
- Improve performance of concurrent task runner — https://github.com/PrefectHQ/prefect/pull/9073
- Improve performance of waiting for task submission — https://github.com/PrefectHQ/prefect/pull/9072
- Add retry on 502 BAD GATEWAY to client — https://github.com/PrefectHQ/prefect/pull/9102
- Update local and remote file systems to return path on write — https://github.com/PrefectHQ/prefect/pull/8965
- Add artifacts `/count` route — https://github.com/PrefectHQ/prefect/pull/9022
- Improve performance of automatic block registration — https://github.com/PrefectHQ/prefect/pull/8838
- Improve performance of log retrieval queries — https://github.com/PrefectHQ/prefect/pull/9035
- Improve performance of artifact retrieval — https://github.com/PrefectHQ/prefect/pull/9061 / https://github.com/PrefectHQ/prefect/pull/9064
- Add `--type` option to create work-pool CLI — https://github.com/PrefectHQ/prefect/pull/8993

> 

Reviewer 3: ```suggestion
- Add flow names to sub flows on the flow run timeline graph - https://github.com/PrefectHQ/prefect-ui-library/pull/1304

```

# Prefect Release Notes

## Release 2.10.0

### Projects

...

- https://github.com/PrefectHQ/prefect/pull/8930
- https://github.com/PrefectHQ/prefect/pull/9103
- https://github.com/PrefectHQ/prefect/pull/9105
- https://github.com/PrefectHQ/prefect/pull/9112
— https://github.com/PrefectHQ/prefect/pull/9093
— https://github.com/PrefectHQ/prefect/pull/9083
- https://github.com/PrefectHQ/prefect/pull/9041

### Variables

...

- https://github.com/PrefectHQ/prefect/pull/9088

### Events

> 

Reviewer 4: ```suggestion
Continuing the rollout of (events[https://docs.prefect.io/concepts/events-and-resources/]) as a primary unit of observability in Prefect Cloud, the client will now by default emit events for all block method calls, which can be seen and configured for use in automations via the Event Feed.

![image](https://user-images.githubusercontent.com/26799928/230421783-997e4fda-a02f-4bf4-88e1-f51a2f890cf5.png)

Configure the prefect client to by default send block method call telemetry to prefect cloud - https://github.com/PrefectHQ/prefect/pull/9107
```

# Prefect Release Notes

## Release 2.10.0

### Projects

...

- https://github.com/PrefectHQ/prefect/pull/8930
- https://github.com/PrefectHQ/prefect/pull/9103
- https://github.com/PrefectHQ/prefect/pull/9105
- https://github.com/PrefectHQ/prefect/pull/9112
— https://github.com/PrefectHQ/prefect/pull/9093
— https://github.com/PrefectHQ/prefect/pull/9083
- https://github.com/PrefectHQ/prefect/pull/9041

### Variables

...

- https://github.com/PrefectHQ/prefect/pull/9088

### Events

> Continuing the rollout of (events[https://docs.prefect.io/concepts/events-and-resources/]) as a primary unit of observability in Prefect Cloud, the client will now by default emit events for all block method calls, which can be seen and configured for use in automations via the Event Feed.

Reviewer 4: ```suggestion
Continuing the rollout of (events[https://docs.prefect.io/concepts/events-and-resources/]) as a primary unit of observability in Prefect Cloud, the Prefect client will now by default emit events for all block method calls. This telemetry can be analyzed in the Event feed, where users can view the interactions their flows and tasks have with external systems such as storage locations, notification services, and infrastructure. Additionally, users can configure automations based on these events, such as when a file is uploaded to a storage location or an infrastructure process or an infrastructure block starts a job.
```
